datafusion/core/src/physical_plan/repartition/sort_preserving_repartition.rs

```diff
@@ -0,0 +1,357 @@
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+//! This file implements an order-preserving repartitioning operator
+//! mapping N input partitions to M output partitions.
+
+use std::pin::Pin;
+use std::sync::Arc;
+use std::task::{Context, Poll};
+use std::{any::Any, vec};
+
+use super::metrics::{ExecutionPlanMetricsSet, MetricsSet};
+use super::SendableRecordBatchStream;
+
+use crate::physical_plan::common::{
+    transpose, AbortOnDropMany, AbortOnDropSingle, SharedMemoryReservation,
+};
+use crate::physical_plan::metrics::BaselineMetrics;
+use crate::physical_plan::repartition::distributor_channels::{
+    partition_aware_channels, DistributionReceiver, DistributionSender,
+};
+use crate::physical_plan::repartition::{
+    pull_from_input, wait_for_task, MaybeBatch, RepartitionMetrics,
+};
+use crate::physical_plan::sorts::streaming_merge;
+use crate::physical_plan::{
+    DisplayFormatType, EquivalenceProperties, ExecutionPlan, Partitioning,
+    RecordBatchStream, Statistics,
+};
+
+use arrow::datatypes::SchemaRef;
+use arrow::record_batch::RecordBatch;
+use datafusion_common::Result;
+use datafusion_execution::memory_pool::MemoryConsumer;
+use datafusion_execution::TaskContext;
+use datafusion_physical_expr::PhysicalSortExpr;
+
+use futures::{FutureExt, Stream};
+use hashbrown::HashMap;
+use log::trace;
+use parking_lot::Mutex;
+use tokio::task::JoinHandle;
+
+type InputPartitionsToCurrentPartitionSender = Vec<DistributionSender<MaybeBatch>>;
+type InputPartitionsToCurrentPartitionReceiver = Vec<DistributionReceiver<MaybeBatch>>;
+
+/// Inner state of [`SortPreservingRepartitionExec`].
+#[derive(Debug)]
+struct SortPreservingRepartitionExecState {
```

Comment:
I can't help noticing that a great deal of this file is almost the same as https://github.com/apache/arrow-datafusion/blob/80d47cce606ab7f22a28d1954aa38a0bda391eda/datafusion/core/src/physical_plan/repartition/mod.rs#L291-L484

This code also has several non trivial parts to it (like memory tracking, yielding back to tokio, tasks. etc) much of which is covered by tests in https://github.com/apache/arrow-datafusion/blob/80d47cce606ab7f22a28d1954aa38a0bda391eda/datafusion/core/src/physical_plan/repartition/mod.rs#L685-L1216

However, since the order aware repartition doesnt' use the same code (it uses a copy) many of those cases are not covered by the existing tests. 

I wonder if there is a way to avoid the copy/paste code and only parameterizing what is different -- specifically I think the construction of the `DistributionSender` / `DistributionReceiver` Vecs and them how to combine the results if each receiver only has a single input.

Is this something you tried?

Reply:
AFAIK @mustafasrepo tried to avoid duplications in general, but didn't engage in non-trivial refactoring to avoid them to maximum extent. Do you have any ideas on how we can refactor to reduce duplication?

Reply:
If there is no easy/obvious way, we can merge this and do another PR for refactoring/deduplication after some more in-depth thinking.

Reply:
> Do you have any ideas on how we can refactor to reduce duplication?

There were two main differences I saw:
1. The order preserving variant has `Vec< DistributionSender>` and `Vec< DistributionReceiver>`  and the normal one has `DistributionSender>` and `DistributionReceiver` 
2. If there were a `Vec< DistributionReceiver>` then they needed to be merged

I wonder if it would be possible to have the repartition operator always have a `Vec< DistributionSender>` and `Vec< DistributionReceiver>` and then special case if there is only one receiver (e.g. don't merge) 🤔

Reply:
> If there is no easy/obvious way, we can merge this and do another PR for refactoring/deduplication after some more in-depth thinking.

I agree

Reply:
@crepererum, what do you think about @alamb's suggestion? Can we apply DRY here? Assuming we can, could that have any negative performance impact on the unordered case, or not?

IIRC you worked on this operator before so maybe you'd have useful insights here.

Reply:
I don't think this DRY would impact perf, at least not enough to have duplicated code. So let's simplify if we can :+1: